### PR TITLE
Mongoose wpool abstraction

### PR DESCRIPTION
**Pull request by kzemek**
This PR addresses #

Proposed changes include:
* describe the functionality changes
* describe new or updated tests
* describe changes to the documentation

### CODE REVIEW

File: src/event_pusher/mod_event_pusher_push.erl [hunk 1]
```diff
@@ -66,8 +66,9 @@
 start(Host, Opts) ->
     ?INFO_MSG("mod_event_pusher_push starting on host ~p", [Host]),
 
-    {ok, _} = wpool:start_sup_pool(gen_mod:get_module_proc(Host, ?MODULE),
-                                   gen_mod:get_opt(wpool, Opts, [])),
+    mongoose_wpool:ensure_started(),
+    WpoolOpts = [{strategy, available_worker} | gen_mod:get_opt(wpool, Opts, [])],
```
Comment (michalwski) [inline src/event_pusher/mod_event_pusher_push.erl]: Do we allow to set the `strategy` in the config file? If so this line will result in selecting `available_worker` strategy even if other is specified in the config.

File: src/mongoose_http_client.erl [hunk 2]
```diff
@@ -44,6 +43,16 @@
 
 -spec start() -> ok.
 start() ->
+    case ets:info(?MODULE) of
+        undefined ->
+            Heir = case whereis(ejabberd_sup) of
```
Comment (michalwski) [inline src/mongoose_http_client.erl]: This trick is already used in other places in the code. It'd be good to unify this before we have too many incarnations of it.

File: src/mongoose_wpool.erl [hunk 3]
```diff
@@ -9,35 +9,135 @@
...
+
+get_pool_settings(Type, Host, Tag) ->
+    case ets:lookup(?MODULE, {Type, Host, Tag}) of
+        [PoolOpts] -> PoolOpts;
+        [] -> undefined
+    end.
 
-delete_pool_settings(PoolName) ->
-    ets:delete(pool_settings_tab(), PoolName).
+get_pools() ->
```
Comment (michalwski) [inline src/mongoose_wpool.erl]: I think it'd be more informative if the function returns more info about the pools than only the name.

**Comment by mongoose-im** [comment]
[5518.1](https://travis-ci.org/esl/MongooseIM/jobs/423888565) / Erlang 19.3 / small_tests / 3bf199c0319f112d9908164c01a3d6cb051dc675
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2060/5518/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2060/5518/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-03_12.41.52/lib.mongooseim.logs/run.2018-09-03_12.41.52/suite.log.html)

File: src/mongoose_wpool.erl (hunk 3 above)
Comment (kzemek) [reply src/mongoose_wpool.erl]: That's what it's supposed to be in the first iteration: https://redmine.erlang-solutions.com/issues/19818

File: src/event_pusher/mod_event_pusher_push.erl (hunk 1 above)
Comment (kzemek) [reply src/event_pusher/mod_event_pusher_push.erl]: I don't believe we do, explicitly

File: src/mongoose_wpool.erl (hunk 3 above)
Comment (michalwski) [reply src/mongoose_wpool.erl]: Ok, fair enough.

File: src/event_pusher/mod_event_pusher_push.erl (hunk 1 above)
Comment (michalwski) [reply src/event_pusher/mod_event_pusher_push.erl]: All right.

File: src/mongoose_http_client.erl (hunk 2 above)
Comment (fenek) [reply src/mongoose_http_client.erl]: There is an internal ticket for this. :)

File: src/mongoose_wpool.erl [hunk 4]
```diff
@@ -9,35 +9,135 @@
...
+
+start(Type, Host, PoolOpts) ->
+    start(Type, Host, default, PoolOpts).
+
+start(Type, Host, Tag, PoolOpts) ->
+    {Opts0, WpoolOpts} = proplists:split(PoolOpts, [strategy, call_timeout]),
+    Opts = lists:append(Opts0),
+    case wpool:start_sup_pool(make_pool_name(Type, Host, Tag), WpoolOpts) of
+        {ok, Pid} ->
+            Strategy = proplists:get_value(strategy, Opts, best_worker),
```
Comment (michalwski) [inline src/mongoose_wpool.erl]: It looks to me that the `strategy` option will land in `WpoolOpts` var as in line [44](https://github.com/esl/MongooseIM/pull/2060/files#diff-293b45de04be8a714405e0a6abdd3048R44). This means that `Opts` var will never have `strategy` option and this also means that the `Strategy` var in this line is always `best_worker`.

What did I miss in my reasoning?

Comment (michalwski) [reply src/mongoose_wpool.erl]: Ah, sorry may bad... It works as expected, ignore my previous comment.

**Comment by mongoose-im** [comment]
[5521.1](https://travis-ci.org/esl/MongooseIM/jobs/423919524) / Erlang 19.3 / small_tests / 051ddd7f7daaa46b7e1be0dc94a1a160896aff9e
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2060/5521/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2060/5521/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-03_14.37.45/lib.mongooseim.logs/run.2018-09-03_14.37.46/suite.log.html)

**Comment by mongoose-im** [comment]
[5535.1](https://travis-ci.org/esl/MongooseIM/jobs/424789781) / Erlang 19.3 / small_tests / f0a96abaab5f8a208d29161b552719bd105845a6
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2060/5535/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2060/5535/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-05_13.22.44/lib.mongooseim.logs/run.2018-09-05_13.22.45/suite.log.html)

**Comment by mongoose-im** [comment]
[5536.1](https://travis-ci.org/esl/MongooseIM/jobs/424803350) / Erlang 19.3 / small_tests / 5b93d07f4fe684f5519f3126d6d3f7bd908d9886
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2060/5536/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2060/5536/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-05_13.53.46/lib.mongooseim.logs/run.2018-09-05_13.53.47/suite.log.html)

**Comment by mongoose-im** [comment]
[5537.1](https://travis-ci.org/esl/MongooseIM/jobs/424830020) / Erlang 19.3 / small_tests / 6b91b401e5319f92439368a9f437107a071e7724
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2060/5537/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2060/5537/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-05_14.53.49/lib.mongooseim.logs/run.2018-09-05_14.53.50/suite.log.html)

File: src/mongoose_wpool.erl [hunk 5]
```diff
@@ -9,35 +9,136 @@
...
+
+get_pool_settings(Type, Host, Tag) ->
+    case ets:lookup(?MODULE, {Type, Host, Tag}) of
+        [PoolOpts] -> PoolOpts;
+        [] -> undefined
+    end.
 
-delete_pool_settings(PoolName) ->
-    ets:delete(pool_settings_tab(), PoolName).
+get_pools() ->
```
Comment (michalwski) [inline src/mongoose_wpool.erl]: I think it be worth to have tests for this function. Currently it's not covered by any test: https://codecov.io/gh/esl/MongooseIM/compare/424c413b04e09227f2c1a16b356250df7c3135ff...ee098b80ec87db0dd6b09964adc127f90f0d5ff0/diff#D10-135

File: src/mongoose_wpool.erl [hunk 6]
```diff
@@ -9,35 +9,136 @@
...
+    lists:map(fun(#mongoose_wpool{name = Name}) -> Name end, ets:tab2list(?MODULE)).
 
-pool_settings_tab() ->
-    mongoose_worker_pool_settings.
+stats(Type, Host, Tag) ->
```
Comment (michalwski) [inline src/mongoose_wpool.erl]: This function is also not covered by tests (https://codecov.io/gh/esl/MongooseIM/compare/424c413b04e09227f2c1a16b356250df7c3135ff...ee098b80ec87db0dd6b09964adc127f90f0d5ff0/diff#D10-138) I think it'd be worth to add a test for it.

**Comment by mongoose-im** [comment]
[5541.1](https://travis-ci.org/esl/MongooseIM/jobs/425199290) / Erlang 19.3 / small_tests / 7fc02c0aac3d49ea29d9aa775677ad80a04e4554
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2060/5541/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2060/5541/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-06_10.33.46/lib.mongooseim.logs/run.2018-09-06_10.33.47/suite.log.html)

**Comment by mongoose-im** [comment]
[5543.1](https://travis-ci.org/esl/MongooseIM/jobs/425203222) / Erlang 19.3 / small_tests / d1e82d40382f1f96baf32bbc113e9958121772cc
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2060/5543/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2060/5543/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-06_10.46.26/lib.mongooseim.logs/run.2018-09-06_10.46.26/suite.log.html)

**Pull request merged**